features/step_definitions/parallel_steps.ts

```diff
@@ -0,0 +1,93 @@
+import { Then } from '../../'
+import { World } from '../support/world'
+import _, { Dictionary } from 'lodash'
+import { messages } from '@cucumber/messages'
+import { expect } from 'chai'
+
+interface TestCaseParser {
+  pickles: Dictionary<messages.IPickle>
+  testCases: Dictionary<string>
+  running: Set<string>
+  testStarts: Dictionary<string>
+}
+
+function defaultHandlers({
+  pickles,
+  testCases,
+  running,
+  testStarts,
+}: TestCaseParser): Record<string, Function> {
+  return {
+    pickle: (p: messages.IPickle) => (pickles[p.id] = p),
+    testCase: (t: messages.TestCase) => (testCases[t.id] = t.pickleId),
+    testCaseStarted: (t: messages.TestCaseStarted) => {
+      running.add(t.testCaseId)
+      testStarts[t.id] = t.testCaseId
+    },
+    testCaseFinished: (t: messages.TestCaseFinished) =>
+      running.delete(testStarts[t.testCaseStartedId]),
+  }
+}
+
+function parse(
+  envelopes: messages.IEnvelope[],
+  handlers: Record<string, Function>
+): void {
+  _.each(envelopes, (envelope) =>
+    _.forOwn(envelope, (value, key) => {
+      ;(handlers[key] || _.noop)(value)
+    })
+  )
+}
+
+function verifyTandem(
```

Comment:
This feels a bit complex to me. Using the handlers and verifying as it parses each envelope while efficient, is a bit hard to follow.

Since our assertion are all on the pickle level, thoughts on having a function that takes in a list of envelopes and returns an array where each element is a list of pickles that are running simultaneously.

Then we can assert that list is empty, or certain things about each list of pickles. I think then could have something like 

```
Then the following pickles run simultaneously:
  | complex1, simple4 |
  | complex1, simple5 |
  | complex2, simple5 |
  | complex2, simple6 |
```
 (To make that more reliable, may need to change time for each (maybe make complex take 350 milliseconds)

Reply:
I don't see a way around most of what I'm doing. Given a list of envelops there seems to be no easy way to determine the simultaneously running pickles. From what I gather, to derive this information we must follow a chain that dives past the pickle level.
```
Pickle
pickleId -> testCase
testCaseId -> testCaseStarted
testCaseStartedId -> testCaseFinished
```
I'm using the `started`  & `finished` states to determine parallel processing. I have yet to discover another option. Granted, I'm not too familiar with cucumberjs inner workings. 

Also, while I agree that simplicity is a necessity I'm not sure it wins out in this case if the test will become less reliable and slower.  

If there is no other way to do this then I can attempt to refactor it to be more clear (maybe through verbosity). Please provide additional direction.

Thanks!

Reply:
I agree you have to dive past the pickle level to actually verify it, I'm just suggesting a different interface so that someone looking at the test can understand what is being verified without looking at the code. 

I don't think things will get slower as we still go through all envelopes once and the extra assertion time to me would be negligible.

Let me try to implement my thoughts.

Reply:
@eman2673 if it's okay I'm going to experiment a bit with the idea of Charlie's above and see where it leads, it should help me understand the code a bit better as well.